File: vocabtivator-sentence/src/main/java/org/bubbasmith/vocabtivator/integration/SentenceServiceImpl.java
package org.bubbasmith.vocabtivator.integration;

import org.bubbasmith.vocabtivator.model.Sentence;
import org.bubbasmith.vocabtivator.integration.dao.SearchStatisticsEntity;
import org.bubbasmith.vocabtivator.integration.dao.SearchStatisticsRepository;
import org.bubbasmith.vocabtivator.integration.dao.SentenceEntity;
import org.bubbasmith.vocabtivator.integration.dao.SentenceRepository;
import org.bubbasmith.vocabtivator.integration.external.ExternalSentencesSearch;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;

@Service
public class SentenceServiceImpl implements SentenceService {

    private final ExternalSentencesSearch externalSentencesSearch;
    private final SentenceRepository sentenceRepository;
    private final SearchStatisticsRepository searchStatisticsRepository;
    private final static Logger LOGGER = LoggerFactory.getLogger(SentenceServiceImpl.class);

    public SentenceServiceImpl(ExternalSentencesSearch externalSentencesSearch,
                               SentenceRepository sentenceRepository,
                               SearchStatisticsRepository searchStatisticsRepository) {
        this.externalSentencesSearch = externalSentencesSearch;
        this.sentenceRepository = sentenceRepository;
        this.searchStatisticsRepository = searchStatisticsRepository;
    }

    @Override
    public Mono<Sentence> getSentenceForWord(String word) {
        return sentenceRepository.findRandomForWord(word).flatMap(randomSentence -> {
            if (randomSentence != null) {
                return incWordUsageCount(word).thenReturn(new Sentence()
                        .setText(randomSentence.getText())
                        .setPlaceholders(randomSentence.getPlaceholders()));
            } else {
                LOGGER.info("Sentence for the word '{}' not found, searching in external resources...", word);
                return Mono.just(externalSentencesSearch.findSentencesForWord(word)).flatMap(sentences -> {
                    List<SentenceEntity> sentenceEntities = sentences.stream()
                            .map(s -> new SentenceEntity()
                                    .setText(s.getText())
                                    .setPlaceholders(s.getPlaceholders())
                                    .setVocabWord(word))
                            .collect(Collectors.toList());

                    return sentenceRepository.saveAll(sentenceEntities).then(incWordUsageCount(word)).thenReturn(sentences.get(new Random().nextInt(sentences.size())));
                });
            }
        });
    }

    private Mono<SearchStatisticsEntity> incWordUsageCount(String word) {
        return searchStatisticsRepository.findByVocabWord(word)
                .or(Mono.just(new SearchStatisticsEntity().setVocabWord(word).setSearchDates(new ArrayList<>())))
                .map(s -> {
                    s.getSearchDates().add(new Date());
                    return s;
                })
                .flatMap(searchStatisticsRepository::save);
    }
}

File: vocabtivator-sentence/src/main/java/org/bubbasmith/vocabtivator/integration/SentenceServiceImpl.kt
package org.bubbasmith.vocabtivator.integration

import kotlinx.coroutines.reactive.awaitFirstOrDefault
import kotlinx.coroutines.reactive.awaitFirstOrNull
import kotlinx.coroutines.reactive.awaitSingle
import org.bubbasmith.vocabtivator.integration.dao.SearchStatisticsEntity
import org.bubbasmith.vocabtivator.integration.dao.SearchStatisticsRepository
import org.bubbasmith.vocabtivator.integration.dao.SentenceEntity
import org.bubbasmith.vocabtivator.integration.dao.SentenceRepository
import org.bubbasmith.vocabtivator.integration.external.ExternalSentencesSearch
import org.bubbasmith.vocabtivator.model.Sentence
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.*

@Service
class SentenceServiceImpl(
        private val externalSentencesSearch: ExternalSentencesSearch,
        private val sentenceRepository: SentenceRepository,
        private val searchStatisticsRepository: SearchStatisticsRepository) : SentenceService {

    private val log = LoggerFactory.getLogger(SentenceServiceImpl::class.java)

    override suspend fun getSentenceForWord(word: String): Sentence {
        val randomSentence = sentenceRepository.findRandomForWord(word).awaitFirstOrNull()
        return if (randomSentence != null) {
            this.incWordUsageCount(word)
            Sentence(randomSentence.text, randomSentence.placeholders)
        } else {
            log.info("Sentence for the word '${word}' not found, searching in external resources...")
            val sentences = externalSentencesSearch.findSentencesForWord(word)
            val sentencesEntities = sentences.map { SentenceEntity(null, it.text, word, it.placeholders) }
            sentenceRepository.saveAll(sentencesEntities).awaitSingle()
            incWordUsageCount(word)
            sentences.random()
        }
    }

    private suspend fun incWordUsageCount(word: String): SearchStatisticsEntity {
        val wordStat = searchStatisticsRepository.findByVocabWord(word)
                .awaitFirstOrDefault(SearchStatisticsEntity(null, word, ArrayList()))
        return searchStatisticsRepository.save(
                SearchStatisticsEntity(wordStat.id, wordStat.vocabWord, wordStat.searchDates + listOf(Date()))
        ).awaitSingle()
    }
}
